sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -258,16 +264,33 @@ class FileStreamSource(
    * equal to `end` and will only request offsets greater than `end` in the future.
    */
   override def commit(end: Offset): Unit = {
-    // No-op for now; FileStreamSource currently garbage-collects files based on timestamp
-    // and the value of the maxFileAge parameter.
+    val logOffset = FileStreamSourceOffset(end).logOffset
+
+    if (sourceOptions.cleanSource != CleanSourceMode.NO_OP) {
```

Comment:
Does this need to be done synchronously during commit? Sounds like you could just submit a clean request to a worker thread and let it deal with it...

Only potential problem I see is if the worker thread starts lagging behind the batches, accumulating work until the VM runs out of memory.

Reply:
That was suggested earlier and we agreed to deal with it from follow-up issue - the PR just has been sitting longer than expected. Could we deal with follow-up issue?

Btw, if we don't guarantee the cleanup and do it with best effort (actually we already do it), it won't matter much to do it with background thread - except the problem you've mentioned. That could be remedied via retaining max amount of paths to cleanup (yes, this is based on "best effort").